MNT Update ruff

DimitriPapadopoulos opened the pull request:
#### What does this implement/fix? Explain your changes.

* Bump ruff from 0.5.1 to ~~0.9.10~~ 0.11.0
* Update comments and URL associated to ruff
 
#### Any other comments?

DimitriPapadopoulos committed:
MNT Bump ruff

The linter really needed an update.

DimitriPapadopoulos committed:
MNT Update ruff rules in preview and URL

adrinjalali commented:
Could you please also check that `pre-commit run --all-files` actually passes on all tests?

DimitriPapadopoulos commented:
Sure, I'll update to ruff 0.11.0 too while I'm at it.

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF046

RUF046 Value being cast to `int` is already an integer

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF040

RUF040 Non-string literal used as assert message

DimitriPapadopoulos commented:
By the way, why not run pre-commit during CI?

adrinjalali commented:
> By the way, why not run pre-commit during CI?

I've been thinking the same. I wouldn't be opposed to that actually.

adrinjalali commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -112,7 +112,7 @@ def test_make_classification_informative_features():
         (2, [3 / 4, 1 / 4], 2),
         (10, [1 / 3] * 3, 10),
-        (int(64), [1], 1),
+        ((64), [1], 1),
```
```suggestion
        (64, [1], 1),
```
?

DimitriPapadopoulos replied:
Indeed, but these are automatic fixes and neither `ruff` nor `black` seem to be catching this. But then see my comment https://github.com/scikit-learn/scikit-learn/pull/30976#discussion_r2001143740. I suggest deferring pre-commit fixes to a different PR any way.

adrinjalali commented on sklearn/externals/_arff.py:
we should exclude everything in `externals` from linters.

DimitriPapadopoulos replied:
Filters are not inconsistent:

https://github.com/scikit-learn/scikit-learn/blob/d2dbcff60e69459e990199b83b413264ab401dcd/pyproject.toml#L112-L125

https://github.com/scikit-learn/scikit-learn/blob/d2dbcff60e69459e990199b83b413264ab401dcd/.pre-commit-config.yaml#L14-L17

The problem is that `pre-commit` will explicitly run `black`, `ruff` and other tools on each individual file, without taking into account the filters in `pryproject.toml`. What happens then depends on whether tools such as `black` or `ruff` skip an explicit file argument if the file is part of its ignore list. Clearly `black` does not:
```console
$ black sklearn/externals/_arff.py
reformatted sklearn/externals/_arff.py

All done! ✨ 🍰 ✨
1 file reformatted.
$ 
```

I suggest deferring pre-commit fixes to a different PR: pre-commit is currently broken and fixing it is unrelated to the initial intent of this PR.

DimitriPapadopoulos commented:
Also, `ruff` and `black` are already run by Azure pipelines. It wouldn't make sense to run these tools both in pre-commit.ci and Azure pipelines.

DimitriPapadopoulos commented:
I have reverted the pre-commit fixes. Please follow-up in #31013.

DimitriPapadopoulos committed:
MNT let ruff infer the min Python

It will infer the Python version from pyproject.toml itself:
	requires-python = ">=3.10"

DimitriPapadopoulos committed:
MNT Bump Python in CircleCI to match min version

DimitriPapadopoulos committed:
Apply ruff rule RUF007

RUF007 Prefer `itertools.pairwise()` over `zip()` when iterating over successive pairs

adrinjalali commented:
> A maintainer should authorize pre-commit.ci as an app in the settings of this repository:
> https://pre-commit.ci/

I meant running `pre-commit` as a command in the CI, not the action. That action is a massive one, and rather not depend on it, and we also won't be using its capability of fixing issues in the PRs anyway.
